services/payments/datastore.go

```diff
@@ -388,3 +331,102 @@ func (s Service) GetTransactionFromDocID(ctx context.Context, docID string) (*Tr
+	}
+	merkleValid, err := revisionValidInTree(ctx, s.sdkClient, result)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify Merkle tree: %w", err)
+	}
+	if !merkleValid {
+		return nil, fmt.Errorf("invalid Merkle tree for record: %#v", result)
+	}
+	return result, nil
+}
+
+// GetTransactionById returns the latest version of a record from QLDB if it exists, after doing all requisite validation
+func (s *Service) GetTransactionById(ctx context.Context, id *uuid.UUID) (*Transaction, error) {
+	data, err := s.datastore.Execute(context.Background(), func(txn qldbdriver.Transaction) (interface{}, error) {
+		entry, err := s.getQLDBObject(ctx, txn, id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get QLDB record: %w", err)
+		}
+		return entry, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to query QLDB: %w", err)
+	}
+	if data == nil {
+		return nil, nil
+	}
+	assertedData, ok := data.(*QLDBPaymentTransitionHistoryEntry)
+	if !ok {
+		return nil, fmt.Errorf("database response was the wrong type: %#v", data)
+	}
+	transaction, err := assertedData.toTransaction()
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert record to Transaction: %w", err)
+	}
+	return transaction, nil
+}
+
+// getTransactionHistory returns a slice of entries representing the entire state history
+// for a given id.
+func getTransactionHistory(txn wrappedQldbTxnAPI, id *uuid.UUID) ([]QLDBPaymentTransitionHistoryEntry, error) {
+	result, err := txn.Execute("SELECT * FROM history(transactions) AS h WHERE h.metadata.id = ?", id)
+	if err != nil {
+		return nil, fmt.Errorf("QLDB transaction failed: %w", err)
+	}
+	var collectedData []QLDBPaymentTransitionHistoryEntry
+	for result.Next(txn) {
+		var data QLDBPaymentTransitionHistoryEntry
+		err := ion.Unmarshal(result.GetCurrentData(), &data)
+		if err != nil {
+			return nil, fmt.Errorf("ion unmarshal failed: %w", err)
+		}
+		collectedData = append(collectedData, data)
+	}
+	if len(collectedData) > 0 {
+		return collectedData, nil
+	}
+	return nil, nil
+}
+
+// WriteTransaction persists an object in a transaction after verifying that its change
+// represents a valid state transition.
+func (s *Service) WriteTransaction(ctx context.Context, transaction *Transaction) (*Transaction, error) {
+	_, err := s.datastore.Execute(ctx, func(txn qldbdriver.Transaction) (interface{}, error) {
+		// Determine if the transaction already exists or if it needs to be initialized. This call will do all necessary
+		// record and history validation if they exist for this record
+		record, err := s.getQLDBObject(ctx, txn, transaction.ID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query QLDB: %w", err)
+		}
+		_, transaction.Signature, err = transaction.SignTransaction(ctx, s.kmsSigningClient, s.kmsSigningKeyId)
+		if err != nil {
+			return nil, fmt.Errorf("failed to sign transaction: %w", err)
+		}
+
+		if record == nil {
+			return txn.Execute("INSERT INTO transactions ?", transaction)
```

Comment:
we need to also write the public key that was used to sign in addition to the signature, so transaction.PublicKey needs to be set prior to insertion as well.